internal/iapl/policy.go

```diff
@@ -0,0 +1,300 @@
+package iapl
+
+import (
+	"fmt"
+
+	"go.infratographer.com/permissions-api/internal/types"
+)
+
+// PolicyDocument represents a partial authorization policy.
+type PolicyDocument struct {
+	ResourceTypes  []ResourceType
+	TypeAliases    []TypeAlias
+	Actions        []Action
+	ActionBindings []ActionBinding
+}
+
+// ResourceType represents a resource type in the authorization policy.
+type ResourceType struct {
+	Name          string
+	IDPrefix      string
+	Relationships []Relationship
+}
+
+// Relationship represents a named relation between two resources.
+type Relationship struct {
+	Relation       string
+	TargetTypeName string
+}
+
+// TypeAlias represents a named alias to multiple concrete resource types.
+type TypeAlias struct {
+	Name              string
+	ResourceTypeNames []string
+}
+
+// Action represents an action that can be taken in an authorization policy.
+type Action struct {
+	Name string
+}
+
+// ActionBinding represents a binding of an action to a resource type.
+type ActionBinding struct {
+	ActionName       string
+	ResourceTypeName string
+	Conditions       []Condition
+}
+
+// Condition represents a necessary condition for performing an action.
+type Condition struct {
+	RoleBinding        *ConditionRoleBinding
+	RelationshipAction *ConditionRelationshipAction
+}
+
+// ConditionRoleBinding represents a condition where a role binding is necessary to perform an action.
+type ConditionRoleBinding struct {
+}
+
+// ConditionRelationshipAction represents a condition where another action must be allowed on a resource
+// along a relation to perform an action.
+type ConditionRelationshipAction struct {
+	Relation   string
+	ActionName string
+}
+
+// Policy represents an authorization policy as defined by IAPL.
+type Policy interface {
+	Validate() error
+	Schema() []types.ResourceType
+}
+
+var _ Policy = &policy{}
+
+type policy struct {
+	rt map[string]ResourceType
+	ta map[string]TypeAlias
+	ac map[string]Action
+	rb map[string]map[string]struct{}
+	p  PolicyDocument
+}
+
+// NewPolicy creates a policy from the given policy document.
+func NewPolicy(p PolicyDocument) Policy {
+	rt := make(map[string]ResourceType, len(p.ResourceTypes))
+	for _, r := range p.ResourceTypes {
+		rt[r.Name] = r
+	}
+
+	ta := make(map[string]TypeAlias, len(p.TypeAliases))
+	for _, t := range p.TypeAliases {
+		ta[t.Name] = t
+	}
+
+	ac := make(map[string]Action, len(p.Actions))
+	for _, a := range p.Actions {
+		ac[a.Name] = a
+	}
+
+	rb := make(map[string]map[string]struct{}, len(p.ResourceTypes))
+	for _, ab := range p.ActionBindings {
+		b, ok := rb[ab.ResourceTypeName]
+		if !ok {
+			b = make(map[string]struct{})
+			rb[ab.ResourceTypeName] = b
+		}
+
+		b[ab.ActionName] = struct{}{}
+	}
+
+	out := policy{
+		rt: rt,
+		ta: ta,
+		ac: ac,
+		rb: rb,
+		p:  p,
+	}
+
+	return &out
+}
+
+func (v *policy) validateTypeAliases() error {
```

Comment:
check for a `nil` policy? here and below?

Reply:
In general I don't like to do nil checks for pointer receivers unless there's a legitimate reason they could be nil. In this case there isn't, so a panic would be the expected result for me if someone tried to call `validateTypeAliases()` on a nil policy.